command/ca/policy/actions/uris.go

```diff
@@ -0,0 +1,122 @@
+package actions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/urfave/cli"
+
+	"github.com/smallstep/cli/command/ca/policy/policycontext"
+	"github.com/smallstep/cli/flags"
+	"github.com/smallstep/cli/internal/command"
+	"github.com/smallstep/cli/utils/cautils"
+)
+
+// Command returns the policy subcommand.
+func URICommand(ctx context.Context) cli.Command {
+
+	return cli.Command{
+		Name:  "uri",
+		Usage: "...",
+		UsageText: `**uri** <domain> [**--remove**]
+[**--provisioner**=<name>] [**--admin-cert**=<file>] [**--admin-key**=<file>]
+[**--admin-provisioner**=<string>] [**--admin-subject**=<string>]
+[**--password-file**=<file>] [**--ca-url**=<uri>] [**--root**=<file>]
+[**--context**=<name>]`,
+		Description: `**uri** command group provides facilities for ...`,
+		Action: command.InjectContext(
+			ctx,
+			uriAction,
+		),
+		Flags: []cli.Flag{
+			provisionerFilterFlag,
+			flags.AdminCert,
+			flags.AdminKey,
+			flags.AdminProvisioner,
+			flags.AdminSubject,
+			flags.PasswordFile,
+			flags.CaURL,
+			flags.Root,
+			flags.Context,
+			cli.BoolFlag{
+				Name:  "remove",
+				Usage: `removes the provided URIs from the policy instead of adding them`,
+			},
+		},
+	}
+}
+
+func uriAction(ctx context.Context) (err error) {
+
+	clictx := command.CLIContextFromContext(ctx)
+
+	args := clictx.Args()
+	if len(args) == 0 {
+		return errors.New("please provide at least one URI")
+	}
+
+	client, err := cautils.NewAdminClient(clictx)
+	if err != nil {
+		return fmt.Errorf("error creating admin client: %w", err)
+	}
+
+	policy, err := retrieveAndInitializePolicy(ctx, client)
+	if err != nil {
+		return fmt.Errorf("error retrieving policy: %w", err)
+	}
+
+	var uris []string
+	switch {
+	case policycontext.HasSSHHostPolicy(ctx):
+		return errors.New("SSH host policy does not support URIs")
+	case policycontext.HasSSHUserPolicy(ctx):
+		return errors.New("SSH user policy does not support URIs")
+	case policycontext.HasX509Policy(ctx):
+		switch {
+		case policycontext.HasAllow(ctx):
+			uris = policy.X509.Allow.Uris
+		case policycontext.HasDeny(ctx):
+			uris = policy.X509.Deny.Uris
+		default:
+			panic(errors.New("no allow nor deny context set"))
+		}
+	default:
+		panic("no SSH nor X.509 context set")
+	}
+
+	if clictx.Bool("remove") {
+		for _, uri := range args {
+			uris = remove(uri, uris)
+		}
+	} else {
+		uris = append(uris, args...)
+	}
```

Comment:
We should validate the uri, it has to be an *url.URL

Reply:
At this time the rules for URIs only check the domain part. URI values that contain a scheme, for example, will not be allowed to be POSTed. They will result in an error like this:

```
{
    "type": "badRequest",
    "detail": "bad request",
    "message": "error updating authority policy: cannot parse permitted URI domain constraint \"https://www.example.com\": URI domain constraint \"https://www.example.com\" contains scheme (not supported yet)"
}
```

I've implemented it like this, so that users won't be surprised that the scheme is not taken into account (at this time) when requesting a certificate. Any scheme will thus be allowed right now. If we would parse it as an URL here and only take the domain part of it in the rules engine, that's probably not as transparent as this. Support for URI schemes was discussed, but was left out of scope for now.